BIP drafts: SwiftSync Specification (#2152)
bitcoin/bips: bip-xxxx-swiftsync.md

limitations. First, although the lifetime of coins demonstrates an empirical distribution, cache misses occur for coins
that are deleted. This creates unnecessary disk I/O and database compaction. Secondly, given the structure of a block,
coins that are spent are indexed by their outpoint. This creates a requirement for clients to maintain a cache to fetch
coin metadata associated with an outpoint. _SwiftSync_ alleviates both of these limitations, allowing for IBD in as fast
as a client can download blocks and verify signatures.

## Specification

### Definitions

- $H$: A hashing function
- $Hintsfile\_{n}$: Defined in BIP ???
- $UTXO\_{n}$: Unspent outputs at block height $n$

_SwiftSync_ builds on a common observation in cryptography, that _verification_ is often orders of magnitude more
performant than _computation_. What a client seeks to verify when performing _SwiftSync_ is that a unspent transaction
output (UTXO) set indeed corresponds to the chain history downloaded from peers.

A key invariant is that a UTXO set state at height $n$ is equivalent to all of the outputs created in the chain history,
less all of the inputs:

> $Outputs - UTXO\_{n} = Inputs$

Reviewer: `Outputs`, `UTXO\_{n}`, and `Inputs` are sets (in the mathematical sense). I feel that `-` is not as prevalent a notation in this case as `\setminus`.

I would have written `$UTXO\_{n} = Outputs \setminus Inputs$` because we are effectively constructing `UTXO\_{n}`, but not the other sets (that are somehow tracked by the aggregators).

Author: The left-hand-side is everything added to the output aggregate and the right-hand-side is everything added to the input aggregate. These are exactly the values we need to assert are equal, and we arrive at the UTXO set as a side effect.

As far as set notation, I didn't even know \setminus was a symbol. For programmers I think a simple `-` is more common and what I find the most legible.

Given this relationship between the two sets, a client uses hints to $UTXO\_{n}$ to verify the chain history they have
received is correct. This document describes the fully-validating client, however this protocol may be easily extended
to assume-valid assumptions.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and
"OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Aggregation

A client must compare $Outputs - UTXO\_{n}$ with $Inputs$ in a succinct way. Rather than say, comparing the lists, a
client may compute two aggregates, and compare the values at the end of the protocol. The _aggregate_ is defined over
the data structure of the elements and a hashing function. It is recommended, but not strictly required, that a client
uses the typical SHA-256 hash, with an additional randomly generated salt value[^2]. The elements of the set a client
compares are _coins_, which are defined as the following:

> $Coin = Creation Height + Coinbase Flag + Output Script + Amount + Outpoint$

Reviewer: We are effectively building bit patterns that will serve as hash preimages, not numbers, so:

```suggestion
$Coin = Creation Height || Coinbase Flag || Output Script || Amount || Outpoint$
```

Also, because of that, serialization of each piece of information should be formally defined, e.g. https://github.com/bitcoin/bips/blob/861e235e93b40e84e86652ef6e80c2f2dbfc1e17/bip-0340.mediawiki?plain=1#L96-L120

To add an element to an aggregate, a client computes the $H(Coin)$ and interprets that number as an unsigned 256-bit
integer. To update the state of the aggregate, a client simply adds this element to the previous state, modulo 256 bits.
Thus, for a set of $Coin$, the _aggregate_ is defined as:

$Agg = H(Coin\_{0}) + ... + H(Coin\_{i})\\mod2^{256}$

### Protocol

Block validation when performing _SwiftSync_ is nearly the same, only with a few additional steps. First, a client
requires a $Hintsfile\_{n}$ and two aggregates, one for block inputs $Agg\_{inputs}$, one for block outputs

> $Agg\_{outputs}$

Reviewer: ```suggestion
requires a $Hintsfile\_{n}$ and initializes two aggregates with the same value, one for block inputs $Agg\_{inputs}$, the other for block outputs
$Agg\_{outputs}$.
```

1. The two aggregators should be initialized with the same value, otherwise validation will fail. Intuitively, they should start as 0, and I feel this should be specified, but I'm not sure if there might be any advantage to use a e.g. random value.
2. Using `requires` for something that comes from the outside world (the hintsfile) and for the aggregators suggest they also come from an external source, but they are managed internally only.

When downloading blocks, _SwiftSync_ clients will do the following:

1. Download the required block undo data defined in BIP ???
1. Using the undo data, validate the block. A non-exhaustive but explicit list of steps include:
   1. Check the coinbase does not over-claim the block subsidy
   1. Check the number of signature operations
   1. Validate all script executions succeed
   1. Validate non-contextual checks like merkle roots and timestamps

>    1. If the block is invalid, disconnect the peer. If the block is valid, continue

Reviewer: ```suggestion
1. Using the undo data, validate the block. If the block is invalid, fail.
```

1. Since the validation list is not exhaustive and since consensus is not formally defined anyhow, better to leave it out.
2. `If the block is valid continue` doesn't specify what to do in the case the block is invalid, it's implicit we should continue...

> 1. For all inputs, except the coinbase, add them to $Agg\_{inputs}$

Reviewer: This is not wrong or insufficient, although it was specially challenging to grok since `Coin` is defined as:

`Coin = Creation Height + Coinbase Flag + **Output Script + Amount** + Outpoint`

I.e. which output script and amount should I use when processing an input?

By doing a manual toy example, I figured out one should "dereference" the outpoint and retrieve data from a previous transaction (which might live in a different block, which might force sequential processing). If this understanding is right, we are back at square one. IIUC, this info is in the undo data, but it is not mentioned here.

Author: I added "using the spent coin data" to make it clear which output script, amount, etc is used

1. For all outputs:
   1. If the output is _unspendable_, continue
   1. Add the output to $UTXO\_{n}$ if the output is in $Hintsfile\_{n}$ and continue

>    1. Otherwise, add the output to $Agg\_{outputs}$

Reviewer: ```suggestion
1. For each output:
   1. If the output is _unspendable_, continue.
   1. If the output is marked unspent in $Hintsfile\_{n}$, add it to $UTXO\_{n}$. Else, add the output to $Agg\_{outputs}$.
```

Also, see the other comment about the `unspendable` definition. I think it should be here, instead.

Author: Added to the `Definitions` section

Notice here that a client does not have to download blocks in any particular order, and may download blocks from
multiple peers at a time. A client then verifies $Agg\_{outputs} = Agg\_{inputs}$ once they have arrived at height

> $n$[^1]. If the verification succeeds, the client accepts $UTXO\_{n}$ as valid. In the failure case, the client rejects

Reviewer: nit: out of order footnote number.

## Extension to Assume Valid

The protocol may be easily extended, and rather simplified, with similar assumptions to _assume valid_. Rather than
hashing the entirety of coin data, a client may take $H(Outpoint)$ and add these results to the aggregates. This removes
the need to download block undo data from peers, and is compatible with the current peer-to-peer protocol.

There are, however, a number of drawbacks with this valid approach. At the end of the protocol, the client must also
check the total monetary value introduced in the system is less than the expected value for the height $n$. An
equivalency check is not possible, as there are instances where coinbase subsidies were under-claimed. Further, the
number of signature operations cannot be checked for input scripts, and the order of transactions cannot be checked as
well.

Finally, the undo data for the block cannot be computed locally with this approach. This poses a data availability
problem, as fully-validating SwiftSync clients will need their connections to serve undo-data.

## Reference Implementation

- [`swiftsync`](https://github.com/2140-dev/swiftsync) Rust implementation

## Acknowledgements

> Kudos to my colleague Ruben Somsen for creating the protocol described in this specification

Reviewer: Since Ruben is a co-author, it's really weird to have this here.

```
  BIP: ?
  Layer: Applications
  Title: SwiftSync initial block download
  Authors: Robert Netzke <bips@2140.dev>, Ruben Somsen <bips@2140.dev>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: BSD-3-Clause
  Requires: BIP ?, BIP ?
```

> # Abstract

Reviewer 2: IIRC, Markdown only allows a single first-level header per document, so you probably want to bump all of the headers up one level.

```suggestion
## Abstract
```

```
  BIP: ?
  Layer: Applications
  Title: SwiftSync initial block download
  Authors: Robert Netzke <bips@2140.dev>, Ruben Somsen <bips@2140.dev>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: BSD-3-Clause
  Requires: BIP ?, BIP ?
  Discussion: https://groups.google.com/g/bitcoindev/c/FpSWUxItXQs/m/pnfjP6rFCgAJ
```
## Abstract

_SwiftSync_ is a protocol to accelerate initial block download (IBD) using existing cryptographic primitives and minimal state. The protocol is comprised of hash aggregate for a set of elements and a "hintsfile" to indicate the spent-ness of outputs in the chain history. Using these hints, clients may perform IBD in parallel, which maximizes the use of existing system resources and shifts the performance limitations to internet quality.
## Motivation

> Initial block download is the first user experience when using Bitcoin software, and, moreover, is a bootstrapping cost for second layer protocols. Improvements to this process benefit end-users and scaling protocols alike. IBD faces two limitations. First, although the lifetime of coins demonstrates an empirical distribution, cache misses occur for coins that are deleted. This creates unnecessary disk I/O and database compaction. Secondly, given the structure of a block, coins that are spent are indexed by their outpoint. This creates a requirement for clients to maintain a cache to fetch coin metadata associated with an outpoint. _SwiftSync_ alleviates both of these limitations, allowing for IBD in as fast as a client can download blocks and verify signatures.

Reviewer 3: I would think there is also efficiency gains in that the blocks need not be downloaded in linear fashion from peers, and also that blocks can be shared with peers before you have downloaded the entire chain state.

Author: This is mentioned in the abstract.
